text/0158-dynamic-span-creation.md

```diff
@@ -0,0 +1,109 @@
+- The approach to dynamically adapting the expressiveness of traces has been touched upon in research papers, including [Mace](https://dl.acm.org/doi/10.1145/2815400.2815415), [Castanheira](https://dl.acm.org/doi/10.1145/3426746.3434058), and [Sambasivan](https://dl.acm.org/doi/10.1145/2987550.2987568).
+
+We hope that as OpenTelemetry continues to evolve and is implemented in a growing set of libraries, that it will be possible for operators to dynamically activate and deactivate trace points without requiring recompilation of applications or restarting services.
+
+This proposal will further benefit applications deployed in infrastructures that already provide tracing at an infrastructure level.
+Such examples are seen in the development of Envoy and various gateways, which already attach trace and span IDs to requests.
+For many situations, the level of detail provided by Envoy is sufficient for correlating activities in microservice applications.
+In this case, OpenTelemetry libraries could operate in a "dormant" state where no additional spans are created by the application instrumentation or any libraries it uses.
+Instead, OpenTelemetry simply forwards tracing headers and performs other activities not related to span creation, such as injecting span contexts into logging libraries.
+During crisis situations, operators can dynamically activate the OpenTelemetry instrumentation to acquire additional detail on problematic services, but only for short periods of time.
+As Castanheira puts it:
+
+> Despite the increasing number of uses for distributed tracing, it remains, in the wild, focused on the more common and general use cases (such as latency profiling for troubleshooting slow requests).
+
+For this reason, he argues, applications are typically instrumented very coarsely, and perhaps only for a few central workflows. However, he proceeds:
+
+> During crises, operators may be willing to accept a decrease in performance in exchange for a larger wealth of information, but cannot act on this intent because of the static tracing instrumentation.
+
+## Explanation
+
+In order for operators to configure the expressiveness of tracing at runtime, they need a way to communicate with the instrumented applications.
+As inspired by Jaeger, this could be implemented by allowing for OpenTelemetry Collectors to inform applications about the desired level of detail, perhaps using polling. We believe that the level of detail (LoD) should be configurable at at least three levels:
+
+- `PROPAGATION_ONLY`: Only propoagate headers, and suppress the creation of all spans. This will work well with service meshes.
+- `EXTERNAL_SPANS`: Propagate headers, and generate additional client, server, consumer, and producer spans. This will enhance the traces generated by service meshes, but only when the performance impact can be justified.
+- `ALL_SPANS`: Propagate headers, and generate spans at any level of detail
```

Comment:
Specifying which spans to generate is one possible dimension of controlling trace verbosity. There are other dimensions, such as:
  * still capturing the spans but changing the level of details recorded about them, e.g. including / excluding numerous performance metrics (CPU cycles, memory allocs, etc.) which could be expensive to acquire
  * there are also opposite use cases that are very interested in perf metrics but not so much in the static span data such as span name and resource attributes, e.g. for evaluating resource utilization of a particular product feature
  * only capturing spans at certain places in the call graph. For example, some workflows might involve a lot of queries to storage systems (100:1 ratio of db spans to service spans is not unusual), and storage spans are both expensive to capture and not always needed. The verbosity controls can either suppress storage spans completely or replace them with client-side summaries.

Reply:
I agree, there are many dimensions to this. We have also experienced workflows with over 600 db spans. While I understand that this information may be of interest to some clients, our industry partners have a different use case. They mostly collect metrics via integrations with platforms such as AWS. For this reason, their use case for tracing is mostly for debugging and performance optimization by visualising and analyzing workflows. Common to both use cases is the need for a dynamic trace verbosity as opposed to a static degree of instrumentation.

I think we will need more feedback on use cases to properly design verbosity controls of general usefulness. Should we keep the discussion to this pull request, or open an issue somewhere?